nmdc_server/crud.py

```diff
@@ -488,3 +489,90 @@ def release_submission_lock(db: Session, submission_id: str):
     submission.locked_by = None  # type: ignore
     db.commit()
+
+
+#############################
+# SUBMISSION ACCESS CONTROL #
+#############################
+read_roles = [
+    models.SubmissionEditorRole.editor,
+    models.SubmissionEditorRole.metadata_contributor,
+    models.SubmissionEditorRole.owner,
+    models.SubmissionEditorRole.viewer,
+]
+
+metadata_edit_roles = [
+    models.SubmissionEditorRole.editor,
+    models.SubmissionEditorRole.metadata_contributor,
+    models.SubmissionEditorRole.owner,
+]
+
+context_edit_roles = [
+    models.SubmissionEditorRole.editor,
+    models.SubmissionEditorRole.owner,
+]
+
+contributors_edit_roles = [
+    models.SubmissionEditorRole.owner,
+]
+
+
+def can_read(db: Session, submission_id: str, user_orcid: str) -> Optional[bool]:
+    role = (
+        db.query(models.SubmissionRole)
+        .filter(
+            and_(
+                models.SubmissionRole.user_orcid == user_orcid,
+                models.SubmissionRole.submission_id == submission_id,
+            )
+        )
+        .first()
+    )
+    submission: Optional[models.SubmissionMetadata] = db.query(models.SubmissionMetadata).get(
+        submission_id
+    )
+    if submission is None:
+        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Submission not found")
+    is_creator = submission.author_orcid == user_orcid
+    return (is_creator or (role and models.SubmissionEditorRole(role.role) in read_roles)) is True
+
+
+def can_edit_all(db: Session, submission_id: str, user_orcid: str) -> Optional[bool]:
+    role = (
+        db.query(models.SubmissionRole)
+        .filter(
+            and_(
+                models.SubmissionRole.user_orcid == user_orcid,
+                models.SubmissionRole.submission_id == submission_id,
+            )
+        )
+        .first()
+    )
+    submission: Optional[models.SubmissionMetadata] = db.query(models.SubmissionMetadata).get(
+        submission_id
+    )
+    if submission is None:
+        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Submission not found")
+    is_creator = submission.author_orcid == user_orcid
+    return (
+        is_creator or (role and models.SubmissionEditorRole(role.role) in contributors_edit_roles)
+    ) is True
+
+
+def get_submissions_for_user(db: Session, user: models.User):
+    """Return all submissions that a user has permission to view."""
+    all_submissions = db.query(models.SubmissionMetadata)
+
+    if user.is_admin:
+        return all_submissions
+
+    # Use a left outer join to include submissions with no roles for the user,
+    # so we can filter on author orcid
+    permitted_submissions = all_submissions.outerjoin(models.SubmissionRole)
+    permitted_submissions = permitted_submissions.filter(
+        or_(
+            models.SubmissionRole.user_orcid == user.orcid,
+            models.SubmissionMetadata.author_orcid == user.orcid,
```

Comment:
So a particular role (author) is stored not in the roles table, but in the submission metadata itself. I can see the positive here of maintaining 1-1 naturally between authors and submissions by it being right in the submissions table. Some possible negatives are more complex logic like this, and people who can edit the entire submission can potentially also edit the author field (maybe in a hacked client?).

Reply:
Author isn't really a role -- it implies the `owner` role.

Its a consequence of existing submissions not having any roles associated with them now. I could possibly modify the migration script to include a data migration that creates an `owner` role for the author of each submission it finds.

I don't think there's a concern about editing the author field from the client, since that information is stored in a column on the `submission_metadata` table and the [endpoint that updates submissions](https://github.com/microbiomedata/nmdc-server/blob/069b63224a3abb732b95ef2f65293f968a660574/nmdc_server/api.py#L548-L577) only ever modifies the `metadata_submission` (jsonb) column.

Reply:
> Author isn't really a role -- it implies the owner role.

Ah ok that makes more sense.

> I could possibly modify the migration script to include a data migration that creates an owner role for the author of each submission it finds.

If you did this, would it make sense to also remove the `SubmissionMetadata.author_orcid` field entirely in the migration?

Reply:
It might. I didn't add that field so I don't know if it used anywhere else. I can take a look at that as more development is done for more submission roles. This PR is already getting to be pretty large so I'm hesitant to remove that right away.